### PR TITLE
Making MSH3 HTTP/3 implementation run again in its cfilter.

### PR DESCRIPTION
**Pull request by icing**
- test 2500, single GET works
- test 2501, single POST stalls
- test 2502, multiple, sequential GETs each use a new connection since MsH3ConnectionGetState(qconn) no longer reports CONNECTED after one GET.

### CODE REVIEW

File: lib/vquic/msh3.c [hunk 1]
```diff
@@ -110,19 +119,22 @@ static CURLcode msh3_data_setup(struct Curl_cfilter *cf,
   struct HTTP *stream = data->req.p.http;
   (void)cf;
 
-  H3BUGF(infof(data, "msh3_data_setup"));
-  stream->recv_buf = malloc(MSH3_REQ_INIT_BUF_LEN);
+  DEBUGASSERT(stream);
   if(!stream->recv_buf) {
```
Comment (nibanks) [inline lib/vquic/msh3.c:123]: Why would it not already be `NULL`?

File: lib/vquic/msh3.c [hunk 2]
```diff
@@ -131,13 +143,18 @@ static bool msh3request_ensure_room(struct HTTP *stream, size_t len)
 {
   uint8_t *new_recv_buf;
   const size_t cur_recv_len = stream->recv_header_len + stream->recv_data_len;
+
   if(cur_recv_len + len > stream->recv_buf_alloc) {
     size_t new_recv_buf_alloc_len = stream->recv_buf_alloc;
     do {
       new_recv_buf_alloc_len <<= 1; /* TODO - handle overflow */
     } while(cur_recv_len + len > new_recv_buf_alloc_len);
+    CF_DEBUGF(fprintf(stderr, "* enlarging buffer to %zu\n",
```
Comment (nibanks) [inline lib/vquic/msh3.c:152]: Why are you sending these all to `stderr`?

File: lib/cf-socket.c [hunk 3]
```diff
@@ -79,7 +79,7 @@
 #include "memdebug.h"
 
 
-#define DEBUG_CF 0
+#define DEBUG_CF 1
```
Comment (nibanks) [inline lib/cf-socket.c:82]: Did you mean to enable this by default?

File: lib/vquic/msh3.c [hunk 4]
```diff
@@ -189,7 +211,8 @@ static void MSH3_CALL msh3_header_received(MSH3_REQUEST *Request,
               (int)Header->ValueLength, Header->Value);
   }
 
-  stream->recv_header_len += total_len - 1; /* don't include null-terminator */
+  stream->recv_header_len += total_len;
```
Comment (nibanks) [inline lib/vquic/msh3.c:214]: Why this change?

File: lib/vquic/msh3.c [hunk 5]
```diff
@@ -244,15 +276,17 @@ static void MSH3_CALL msh3_complete(MSH3_REQUEST *Request, void *IfContext,
 
 static void MSH3_CALL msh3_shutdown(MSH3_REQUEST *Request, void *IfContext)
 {
-  struct HTTP *stream = IfContext;
+  struct Curl_easy *data = IfContext;
+  struct HTTP *stream = data->req.p.http;
   (void)Request;
   (void)stream;
```
Comment (nibanks) [inline lib/vquic/msh3.c:282]: I wonder if it's not simply better to do:
```
  (void)Request;
  (void)IfContext;
```

File: lib/vquic/msh3.c [hunk 6]
```diff
@@ -446,9 +525,23 @@ static CURLcode cf_connect_start(struct Curl_cfilter *cf,
                                  struct Curl_easy *data)
 {
   struct cf_msh3_ctx *ctx = cf->ctx;
-  bool insecure = !cf->conn->ssl_config.verifypeer;
+  bool verify = !!cf->conn->ssl_config.verifypeer;
+
+  if(verify && (cf->conn->ssl_config.CAfile || cf->conn->ssl_config.CApath)) {
+    /* TODO: find out if there is a way to provide trust anchors to MSH3 */
```
Comment (nibanks) [inline lib/vquic/msh3.c:531]: The latest MsQuic does support this. We can wire it up through msh3 eventually.

File: lib/vquic/msh3.c [hunk 7]
```diff
@@ -512,14 +622,43 @@ static CURLcode cf_msh3_connect(struct Curl_cfilter *cf,
...
+  CF_DEBUGF(infof(data, CFMSG(cf, "clearing ctx")));
   if(ctx) {
     if(ctx->qconn)
       MsH3ConnectionClose(ctx->qconn);
     if(ctx->api)
       MsH3ApiClose(ctx->api);
+
+    if(ctx->active) {
+      /* We share our socket at cf->conn->sock[cf->sockindex] when active.
+       * If it is no longer there, someone has stolen (and hopefully
```
Comment (nibanks) [inline lib/vquic/msh3.c:639]: "someone has stolen [it]"?? Could you explain this some?

File: lib/vquic/msh3.c [hunk 8]
```diff
@@ -512,14 +622,43 @@ static CURLcode cf_msh3_connect(struct Curl_cfilter *cf,
...
+       * closed it) and we just forget about it.
```
Comment (nibanks) [inline lib/vquic/msh3.c:640]:
```suggestion
       * If it is no longer there, someone has stolen (and hopefully
       * closed) it and we just forget about it.
```

File: lib/vquic/msh3.c [hunk 9]
```diff
@@ -529,7 +668,7 @@ static void cf_msh3_close(struct Curl_cfilter *cf, struct Curl_easy *data)
 
   (void)data;
   if(ctx) {
-    cf_msh3_ctx_clear(ctx);
+    cf_msh3_ctx_clear(cf, data);
```
Comment (nibanks) [inline lib/vquic/msh3.c:671]: No need to check `if(ctx)` since `cf_msh3_ctx_clear` already handles that. But really, this function should simply be merged with `cf_msh3_ctx_clear`. No need to have separate ones, IMO.

File: lib/vquic/msh3.c (hunk 1 above)
Comment (icing) [reply lib/vquic/msh3.c:123]: There is another path via `DATA_SETUP` event that might already initialize this.

File: lib/vquic/msh3.c (hunk 2 above)
Comment (icing) [reply lib/vquic/msh3.c:152]: stderr is easier for debugging than stdout and Curl's own `infof()` cannot be called, since this is on the "wrong" thread.

File: lib/vquic/msh3.c [hunk 10]
```diff
@@ -538,11 +677,44 @@ static void cf_msh3_destroy(struct Curl_cfilter *cf, struct Curl_easy *data)
...
 
+static CURLcode cf_msh3_query(struct Curl_cfilter *cf,
+                              struct Curl_easy *data,
+                              int query, int *pres1, void **pres2)
+{
+  struct cf_msh3_ctx *ctx = cf->ctx;
+
+  switch(query) {
+  case CF_QUERY_MAX_CONCURRENT: {
+    /* TODO: we do not have access to this so far, fake it */
```
Comment (nibanks) [inline lib/vquic/msh3.c:693]: You're asking how many parallel requests the peer is currently allowing you to open? Why do you need that? msh3/msquic supports queueing requests until the peer is ready, so the app doesn't actually need to know how many instantaneous requests might be allowed (this value can change at any point in time).

File: lib/cf-socket.c (hunk 3 above)
Comment (icing) [reply lib/cf-socket.c:82]: Leftover from my debugging. Should change that.

File: lib/vquic/msh3.c (hunk 4 above)
Comment (icing) [reply lib/vquic/msh3.c:214]: The lengths were off by one in my testing and the header were not correctly parsed.

File: lib/vquic/msh3.c (hunk 1 above)
Comment (nibanks) [reply lib/vquic/msh3.c:123]: Could you explain the state machine a little?

File: lib/vquic/msh3.c (hunk 7 above)
Comment (icing) [reply lib/vquic/msh3.c:639]: Ideally, the connection filters that create sockets should also be the ones to close them. However, there are code paths on curl that access the `conn->sock[index]` ones and seem to do that. I need to track those cases down.

File: lib/vquic/msh3.c (hunk 2 above)
Comment (nibanks) [reply lib/vquic/msh3.c:152]: Ok.

File: lib/vquic/msh3.c (hunk 4 above)
Comment (nibanks) [reply lib/vquic/msh3.c:214]: Ok. I guess we need some tests. 😄

File: lib/vquic/msh3.c (hunk 7 above)
Comment (nibanks) [reply lib/vquic/msh3.c:639]: That doesn't sound good. If we do expose the socket from MsQuic up to curl and it closes it, I expect really bad things would happen.

File: lib/vquic/msh3.c (hunk 10 above)
Comment (bagder) [reply lib/vquic/msh3.c:693]: If the conditions are right, libcurl adds transfers to an existing connection if it can and multiplexes them. Otherwise it can create a new connection for it. For this to work, it needs to know if the new transfer can be put on the existing connection.

**Review by nibanks — COMMENTED** [review]
LGTM. Once it's in, I'll merge/rebase my PR.

**Comment by bagder** [comment]
Thanks!
